classes/local/event_processor.php

```diff
@@ -0,0 +1,173 @@
+        return $uniquedataflows;
+    }
+
+    /**
+     * Gets all events queued for a specific dataflow.
+     * @param int $dataflowid
+     * @return array array of events for the given dataflow
+     */
+    public static function get_events_for_dataflow(int $dataflowid): array {
+        global $DB;
+        return $DB->get_records(self::TABLE, ['dataflowid' => $dataflowid]);
+    }
+
+    /**
+     * Listens to Moodle events and processes the corresponding event triggers.
+     * @param \core\event\base $event
+     */
+    public static function process_event(\core\event\base $event): void {
+        // Get any triggers listening to this event.
+        $eventdata = $event->get_data();
+        $triggersteps = self::get_listening_triggers($eventdata['eventname']);
+
+        // Trigger each workflow depending on its settings.
+        foreach ($triggersteps as $step) {
+            $dataflow = $step->get_dataflow();
+
+            // Only collect events for enabled dataflows.
+            if (!$dataflow->enabled) {
+                continue;
+            }
+
+            $executionpolicy = $step->get_redacted_config()->executionpolicy;
+
+            switch($executionpolicy) {
+                case self::EXECUTE_IMMEDIATELY:
+                    // Record the triggered event to save the event data.
+                    self::add_triggered_event($dataflow->id, $step->id, json_encode($eventdata));
+
+                    // Execute immediately within this thread.
+                    \tool_dataflows\task\process_dataflows::execute_dataflow($dataflow->id);
+
+                    break;
+                case self::EXECUTE_ADHOC:
+                    // Record the triggered event to save the event data.
+                    self::add_triggered_event($dataflow->id, $step->id, json_encode($eventdata));
+
+                    // Create adhoc task to execute.
+                    $record = [ // TODO refactor ?
+                        'dataflowid' => $dataflow->id
+                    ];
+                    \tool_dataflows\task\process_dataflows::create_adhoc_task($record);
+
+                    break;
+                case self::EXECUTE_ADHOCQUEUED:
+                    // Only record that this event was triggered.
+                    // The next CRON call will then execute the event.
+                    self::add_triggered_event($dataflow->id, $step->id, json_encode($eventdata));
+                    break;
+                default:
+                    throw new \coding_exception("Unknown execution policy");
+            }
+        }
+
+        return;
+    }
+
+    /**
+     * Finds trigger steps that are listening to this event.
+     * @param string $eventname
+     * @return array array of steps
+     */
+    private static function get_listening_triggers(string $eventname): array {
+        // Get all event triggers.
+        global $DB;
```

Comment:
```suggestion
        global $DB;
        // Get all event triggers.
```
Just flipping the order. Globals have consistently stuck to the top, and they generally don't have a comment prior to declaration.